nexus/db-queries/src/db/datastore/volume.rs

```diff
@@ -848,41 +850,82 @@ impl DataStore {
 
         Ok(crucible_targets)
     }
+}
 
-    /// Decrease the usage count for Crucible resources according to the
-    /// contents of the volume. Call this when deleting a volume (but before the
-    /// volume record has been hard deleted).
-    ///
-    /// Returns a list of Crucible resources to clean up, and soft-deletes the
-    /// volume. Note this function must be idempotent, it is called from a saga
-    /// node.
-    pub async fn decrease_crucible_resource_count_and_soft_delete_volume(
-        &self,
+#[derive(Debug, thiserror::Error)]
+enum SoftDeleteTransactionError {
+    #[error("Serde error decreasing Crucible resources: {0}")]
+    SerdeError(#[from] serde_json::Error),
+
+    #[error("Updated {0} database rows, expected 1")]
+    UnexpectedDatabaseUpdate(usize),
+}
+
+impl DataStore {
+    // See comment for `soft_delete_volume`
+    async fn soft_delete_volume_txn(
+        conn: &async_bb8_diesel::Connection<DbConnection>,
         volume_id: Uuid,
-    ) -> Result<CrucibleResources, Error> {
-        // Grab all the targets that the volume construction request references.
-        // Do this outside the transaction, as the data inside volume doesn't
-        // change and this would simply add to the transaction time.
-        let crucible_targets = {
-            let volume =
-                if let Some(volume) = self.volume_get(volume_id).await? {
-                    volume
-                } else {
-                    // The volume was hard-deleted, return an empty
-                    // CrucibleResources
-                    return Ok(CrucibleResources::V1(
-                        CrucibleResourcesV1::default(),
-                    ));
+        err: OptionalError<SoftDeleteTransactionError>,
+    ) -> Result<CrucibleResources, diesel::result::Error> {
+        // Grab the volume, and check if the volume was already soft-deleted.
+        // We have to guard against the case where this function is called
+        // multiple times, and that is done by soft-deleting the volume during
+        // the transaction, and returning the previously serialized list of
+        // resources to clean up.
+        let volume = {
+            use db::schema::volume::dsl;
+
+            let volume = dsl::volume
+                .filter(dsl::id.eq(volume_id))
+                .select(Volume::as_select())
+                .get_result_async(conn)
+                .await
+                .optional()?;
+
+            let volume = if let Some(v) = volume {
+                v
+            } else {
+                // The volume was hard-deleted
+                return Ok(CrucibleResources::V1(
+                    CrucibleResourcesV1::default(),
```

Comment:
Nitpick: Not sure this matters, but why return the oldest format here?

Reply:
Yeah, I don't know, I imagine this was a holdover from the first time this code was written. If nothing else this helps us maintain the condition of "must always be able to deserialize this struct" :)